infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/environment/compose/ComposeInternalEnvironment.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.environment.compose;
+
+import static org.eclipse.che.workspace.infrastructure.docker.ArgumentsValidator.checkNotNull;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import java.io.IOException;
+import java.util.Map;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.installer.server.InstallerRegistry;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.InternalMachineConfig;
+import org.eclipse.che.api.workspace.server.spi.RecipeRetriever;
+import org.eclipse.che.workspace.infrastructure.docker.environment.MachineNormalizer;
+import org.eclipse.che.workspace.infrastructure.docker.environment.compose.model.ComposeEnvironment;
+import org.eclipse.che.workspace.infrastructure.docker.environment.compose.model.ComposeService;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerBuildContext;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerEnvironment;
+
+/**
+ * @author Alexander Garagatyi
+ * @author Alexander Andrienko
+ */
+public class ComposeInternalEnvironment extends InternalEnvironment {
```

Comment:
I think that main idea of this PR is to provide an ability to use parsed environments with different infrastructures. Here compose environment factory produces Docker infra's own DockerEnvironment instead of ComposeEnvironment which is native to the environment of type 'compose'. Can you elaborate on how this fits into redesigned SPI?

Reply:
This reused of what you initially did (to not to produce a lot of changes at ones and make it discussion happen). I agree with you it seems no needs to make the extra step of transforming ComposeEnv to DockerEnv